### PR TITLE
Lr finder improvements

### PR DESCRIPTION
**Pull request by dfalbel**
This PR does a few changes to Lr finder and adds an article explaining how to use it.

The changes are:
- We now preserve the batch size when using lr_finder
- We now support passing datasets directly without needing to transform them into dataloaders
- The plot has been slightly modified.

### CODE REVIEW

File: vignettes/articles/lr-finder.Rmd [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+loss can explode and you might never be able to minimize the loss.
+
+The learning rate can be considered another hyperparameter of your model that
+needs to be tuned but, there are techniques that allow you to select a good
+learning rate for your model without having to use the costly strategy of 
+fitting many models with different learning rates and then choosing the one with 
+better results. 
+
+This [article](https://arxiv.org/abs/1506.01186) by Leslie Smith that became popular
+after their approach being implemented in the popular FastAI framework, proposes
```
Comment (skeydan) [inline vignettes/articles/lr-finder.Rmd:33]:
```suggestion
once their approach had been implemented in the popular FastAI framework, proposes
```

File: vignettes/articles/lr-finder.Rmd [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+that we should start with a very small learning rate and slowly increase it until 
+we reach a high learnig rate. At each iteration we record the loss value and in
```
Comment (skeydan) [inline vignettes/articles/lr-finder.Rmd:35]:
```suggestion
we reach a high learning rate. At each iteration we record the loss value and in
```

File: vignettes/articles/lr-finder.Rmd [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+the end we plot it against the learning rate. We can then use this results to decide
```
Comment (skeydan) [inline vignettes/articles/lr-finder.Rmd:36]:
```suggestion
the end we plot it against the learning rate. We can then use these results to decide
```

File: vignettes/articles/lr-finder.Rmd [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+a good learning rate. That's what `lr_finder` does and we will show how to use it.
```
Comment (skeydan) [inline vignettes/articles/lr-finder.Rmd:37]:
```suggestion
on a good learning rate. That's what `lr_finder` does, and we will show how to use it.
```

File: vignettes/articles/lr-finder.Rmd [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+dir <- "~/Downloads/mnist" # caching directory
+
+train_ds <- mnist_dataset(
+  dir,
+  download = TRUE,
+  transform = transform_to_tensor
+)
+```
+
+We can now define our model. We are going to use a small primitive CNN in the LeNet
```
Comment (skeydan) [inline vignettes/articles/lr-finder.Rmd:51]:
```suggestion
We can now define our model. We are going to use a small, straightforward CNN in the LeNet
```

File: vignettes/articles/lr-finder.Rmd [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  forward = function(x) {
+    x %>% 
+      self$features() %>% 
+      self$classifier()
+  }
+)
+```
+
+We can now use the `lr_finder` function to record the loss with different learning rates.
+It's important use the learning rate process with all other hyperparameters of the
```
Comment (skeydan) [inline vignettes/articles/lr-finder.Rmd:83]:
```suggestion
It's important to use the learning rate finder with all other hyperparameters of the
```

File: vignettes/articles/lr-finder.Rmd [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+model fixed because they can influence the choice of the learning rate. For example,
+depending on the batch size, you might want to chose different learning rates.
```
Comment (skeydan) [inline vignettes/articles/lr-finder.Rmd:85]:
```suggestion
depending on the batch size, you might want to choose different learning rates.
```

File: vignettes/articles/lr-finder.Rmd [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+  loss = torch::nn_cross_entropy_loss(),
+  optimizer = torch::optim_adam
+)
+
+records <- lr_finder(
+  object = model, 
+  data = train_ds, 
+  verbose = FALSE,
+  dataloader_options = list(batch_size = 32),
+  start_lr = 1e-6, # the smalles value that will be experimented
```
Comment (skeydan) [inline vignettes/articles/lr-finder.Rmd:98]:
```suggestion
  start_lr = 1e-6, # the smallest value that will be tried
```

File: vignettes/articles/lr-finder.Rmd [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+  end_lr = 1 # the largest value to be experimented
```
Comment (skeydan) [inline vignettes/articles/lr-finder.Rmd:99]:
```suggestion
  end_lr = 1 # the largest value to be experimented with
```

File: vignettes/articles/lr-finder.Rmd [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+)
+
+str(records)
+```
+
+The result is a data frame with the losses and the learning rate in each step.
+You can use the built in plot method to display the results along with a exponentially
```
Comment (skeydan) [inline vignettes/articles/lr-finder.Rmd:106]:
```suggestion
You can use the built-in plot method to display the exact results, along with a exponentially
```

File: vignettes/articles/lr-finder.Rmd [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+smoothed value of the loss.
+
+```{r}
+plot(records) +
+  ggplot2::coord_cartesian(ylim = c(NA, 5))
+```
+We can see that with small learning rates the loss doesn't decrease. At some point
+the loss starts decreasing until it reaches a point where the loss starts increasing
```
Comment (skeydan) [inline vignettes/articles/lr-finder.Rmd:114]:
```suggestion
the loss starts decreasing until it reaches a point where it starts increasing
```

**Review by skeydan — APPROVED** [review]
Looking good!

**Comment by dfalbel** [comment]
Thanks very much @skeydan !
